testlapack/dgehrd.go

```diff
@@ -0,0 +1,204 @@
+// Copyright ©2016 The gonum Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style
+// license that can be found in the LICENSE file.
+
+package testlapack
+
+import (
+	"fmt"
+	"math"
+	"math/rand"
+	"testing"
+
+	"github.com/gonum/blas"
+	"github.com/gonum/blas/blas64"
+)
+
+type Dgehrder interface {
+	Dgehrd(n, ilo, ihi int, a []float64, lda int, tau, work []float64, lwork int)
+}
+
+func DgehrdTest(t *testing.T, impl Dgehrder) {
+	rnd := rand.New(rand.NewSource(1))
+
+	// Randomized tests for small matrix sizes that will most likely
+	// use the unblocked algorithm.
+	for _, n := range []int{2, 3, 4, 5, 10, 34} {
```

Comment:
Not sure. Fortran says that tau is an array of size n-1 but also that n>=0.
It does not promise that tau is not referenced if n<=1 but it makes sense
to assume so. clapack.Dgehrd references tau always by taking the address of
its first element. One option is to pass nil from clapack to the C call if
the length of tau is zero. Another option is to pass a dummy slice from
cgo.Dgehrd to clapack in such situation. The first option is probably the
correct one but @btracey should confirm it.
On Jun 2, 2016 17:15, "Dan Kortschak" <notifications@github.com> wrote:

> In testlapack/dgehrd.go
> <https://github.com/gonum/lapack/pull/134#discussion_r65500292>:
>
> > +	"testing"
> > +
> > +	"github.com/gonum/blas"
> > +	"github.com/gonum/blas/blas64"
> > +)
> > +
> > +type Dgehrder interface {
> > +	Dgehrd(n, ilo, ihi int, a []float64, lda int, tau, work []float64, lwork int)
> > +}
> > +
> > +func DgehrdTest(t *testing.T, impl Dgehrder) {
> > +	rnd := rand.New(rand.NewSource(1))
> > +
> > +	// Randomized tests for small matrix sizes that will most likely
> > +	// use the unblocked algorithm.
> > +	for _, n := range []int{2, 3, 4, 5, 10, 34} {
>
> Is there anything I need to do there?
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/gonum/lapack/pull/134/files/e9a3ec7ea845313d8223d5f545075f31a13d2638#r65500292>,
> or mute the thread
> <https://github.com/notifications/unsubscribe/ABwJvpruKbiKxbt32yrk0oF1ub3Bh0tEks5qHpEZgaJpZM4IsP3a>
> .
>